### PR TITLE
[Transformations][PagedAttention] Replace attention_mask used for broadcasting only

### PR DESCRIPTION
**Pull request by mryzhov**
### Details:
* Added the `AttentionMaskShapeReplacer` transformation, which replaces subgraphs deriving shape information from `attention_mask` with equivalent queries on a specified input source (e.g., `input_ids` or `inputs_embeds`). This ensures models remain functional when `attention_mask` is removed in PagedAttention mode.

This is a safe, short-term fix intended to quickly address the issue.
It is a temporary solution and will be replaced by the more general approach introduced in [PR #36756](https://github.com/openvinotoolkit/openvino/pull/36756?utm_source=chatgpt.com)

### Tickets:
 - *[CVS-189844](https://jira.devtools.intel.com/browse/CVS-189844)*

### AI Assistance:
 - *AI assistance used: no / yes*
 - *If yes, summarize how AI was used and what human validation was performed (build/tests/manual checks).*

### CODE REVIEW

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+namespace v8 = ov::op::v8;
+
+ov::pass::AttentionMaskShapeReplacer::AttentionMaskShapeReplacer(const Output<Node>& input_source) {
+    MATCHER_SCOPE(AttentionMaskShapeReplacer);
+
+    auto attn_mask = wrap_type<v0::Parameter>(has_static_rank() && [](const Output<Node>& output) {
+        return output.get_names().count("attention_mask") > 0;
+    });
+    auto shape_of = wrap_type<v3::ShapeOf>({attn_mask});
+    auto gather = wrap_type<v8::Gather>({shape_of, any_input(), any_input()});
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp]: I think we should use ShapeOf base and Gather base

Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp]: thanks, fixed

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+
+using ov::pass::pattern::any_input;
+using ov::pass::pattern::has_static_rank;
+using ov::pass::pattern::wrap_type;
+
+namespace v0 = ov::op::v0;
+namespace v3 = ov::op::v3;
+
+ov::pass::AttentionMaskShapeReplacer::AttentionMaskShapeReplacer(const Output<Node>& input_source) {
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:31]: Just asking, are any checks for `input_source` required?

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        // and the input source regardless of their ranks, so the rewrite is limited to it.
+        // This also keeps the same Gather index valid and avoids negative-index ambiguity.
+        for (int64_t index : indices_const->cast_vector<int64_t>()) {
+            if (index != 0) {
+                return false;
+            }
+        }
+
+        const auto old_shape_of = pattern_map.at(shape_of).get_node_shared_ptr();
+        auto new_shape_of = std::make_shared<v3::ShapeOf>(input_source, old_shape_of->get_output_element_type(0));
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:87]: Do we really need to create a new `ShapeOf` op? Can we just replace its 0th output with the new output?

File: src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    auto result = std::make_shared<v0::Result>(gather);
+    return std::make_shared<ov::Model>(ResultVector{result}, ParameterVector{attention_mask, input_source});
+}
+
+}  // namespace
+
+class AttentionMaskShapeReplacerTest : public ::TransformationTestsF {};
+
+TEST_F(AttentionMaskShapeReplacerTest, ReplacesWithInputIdsBatchDim) {
+    auto attention_mask = make_attention_mask();
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp:67]: In _src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp_ there's a'make_param()' function which does the same (but return ov::Node). Consider using it by moving to common utils. Would require you using `nodes_to_params` (the same file). Just a suggestion to avoid code duplication.

File: src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
+// Copyright (C) 2018-2026 Intel Corporation
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp:1]: I'd generally recommend moving the entire code of this file into `src/common/transformations/tests/op_conversions/sdpa_to_paged_attention_test.cpp` We test all the SDPAToPA-related transformations there.

Comment (mryzhov) [reply src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp:1]: We will revert this PR, as it does not make sense to include these tests in the common test suite.

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp (hunk 3 above)
Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:87]: I'd keep the new ShapeOf. Replacing the existing ShapeOf(attention_mask) output outright (replace_node / output().replace()) rewires every consumer of that ShapeOf, but the pass only validates this Gather (axis 0, index 0). Other consumers of the same ShapeOf may read different indices.

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+        // The leading dimensions (batch, sequence) of attention_mask coincide with the leading
+        // dimensions of the input source. Every requested index must be valid for both ranks so
+        // that the same Gather indices keep pointing at the same dimensions after rewiring.
+        for (int64_t index : indices_const->cast_vector<int64_t>()) {
+            const int64_t normalized_index = ov::util::normalize(index, attn_mask_rank);
+            if (normalized_index < 0 || normalized_index >= source_rank) {
+                return false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp]: fixed

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace v0 = ov::op::v0;
+namespace v3 = ov::op::v3;
+
+ov::pass::AttentionMaskShapeReplacer::AttentionMaskShapeReplacer(const Output<Node>& input_source) {
+    MATCHER_SCOPE(AttentionMaskShapeReplacer);
+
+    auto attn_mask = wrap_type<v0::Parameter>(has_static_rank() && [](const Output<Node>& output) {
+        return output.get_names().count("attention_mask") > 0;
+    });
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:36]: fixed

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    auto gather = wrap_type<ov::op::util::GatherBase>({shape_of, any_input(), any_input()});
+
+    ov::matcher_pass_callback callback = [=](ov::pass::pattern::Matcher& m) {
+        const auto& pattern_map = m.get_pattern_value_map();
+
+        auto gather_node = pattern_map.at(gather).get_node_shared_ptr();
+        const auto indices_const = ov::util::get_constant_from_source(gather_node->input_value(1));
+        if (!indices_const) {
+            return false;
+        }
```
[the remark this answers is not in the document]
Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:63]: fixed

File: src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp (hunk 4 above)
Comment (mryzhov) [reply src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp:67]: good point, but those helpers are local (anonymous-namespace) to the SDPA test file. To reuse them I need to move them into a shared header.

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp (hunk 2 above)
Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:31]: sure, thanks, added

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+namespace v0 = ov::op::v0;
+namespace v3 = ov::op::v3;
+
+ov::pass::AttentionMaskShapeReplacer::AttentionMaskShapeReplacer(const Output<Node>& input_source) {
+    MATCHER_SCOPE(AttentionMaskShapeReplacer);
+
+    auto attn_mask = wrap_type<v0::Parameter>(has_static_rank() && [](const Output<Node>& output) {
+        return output.get_names().count("attention_mask") > 0;
+    });
+    auto shape_of = wrap_type<ov::op::util::ShapeOfBase>({attn_mask});
```
Comment (evkotov) [inline src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:37]: The matcher triggers on any ShapeOf(attention_mask) -> Gather chain and does not verify that the Gather result actually feeds a broadcasting op (Broadcast/Tile). The PR title and the docstring say "used for broadcasting only", but the implementation rewires any batch-dimension query. Given this is an explicitly temporary fix this is likely acceptable, but the scope is broader than the name suggests — please either narrow the pattern to broadcast consumers or reword the title/comment to match the actual behavior.

**Review by evkotov — APPROVED** [review]
The test helper build_model hardcodes axis = {0}, so several guard branches are never exercised: acceptance of axis == -1 and rejection of axis != 0 (.cpp:56-60), rejection of a dynamic-rank / rank-0 input_source (.cpp:39-42), and rejection of non-constant indices/axis (.cpp:45-47, 52-54). Please add negative tests covering these, and parametrize the Gather axis in build_model.

File: src/common/transformations/include/transformations/paged_attention/attention_mask_shape_replacer.hpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * Some models query the batch dimension from the attention_mask shape.
+ * In PagedAttention mode the attention_mask parameter is removed, so these shape
+ * queries must be rewired to an always-present input (input_ids or inputs_embeds).
+ * The attention_mask batch dimension (index 0) coincides with the batch dimension
+ * of input_ids / inputs_embeds, so the same Gather index remains valid.
+ *
+ * Before:
+ *
+ *  ┌─────────────────┐
```
Comment (CuriousPanCake) [inline src/common/transformations/include/transformations/paged_attention/attention_mask_shape_replacer.hpp:32]: Adjust the ASCII to the recent Gemma4-specific pattern changes.

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    auto gather = wrap_type<ov::op::util::GatherBase>({shape_of, any_input(), any_input()});
+    auto concat = wrap_type<v0::Concat>({gather, any_input(), any_input()});
+    auto broadcast = wrap_type<v3::Broadcast>({any_input(), concat});
+    auto position_ids = wrap_type<v0::Parameter>(has_static_rank() && [](const Output<Node>& output) {
+        return output.get_names().count("position_ids") > 0;
+    });
+    auto position_ids_unsqueeze_0 = ov::pass::pattern::optional<v0::Unsqueeze>({position_ids, any_input()});
+    auto position_ids_unsqueeze_1 = ov::pass::pattern::optional<v0::Unsqueeze>({position_ids_unsqueeze_0, any_input()});
```
Comment (CuriousPanCake) [inline src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:45]: 2 Unsqueeze ops one after another?

File: src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+TEST_F(AttentionMaskShapeReplacerTest, ReplacesWithInputsEmbedsBatchDim) {
+    auto attention_mask = make_attention_mask();
+    auto inputs_embeds = std::make_shared<v0::Parameter>(element::f32, PartialShape{-1, -1, 2048});
+    inputs_embeds->set_friendly_name("inputs_embeds");
+    model = build_model(attention_mask, attention_mask, inputs_embeds, {0});
+
+    manager.register_pass<ov::pass::AttentionMaskShapeReplacer>(inputs_embeds);
+
+    auto attention_mask_ref = make_attention_mask();
+    auto inputs_embeds_ref = std::make_shared<v0::Parameter>(element::f32, PartialShape{-1, -1, 2048});
```
Comment (CuriousPanCake) [inline src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp:98]: Are you sure inputs_embeds_ref can be 3D at this point? I guess, this is the reason the test is failing.

Comment (mryzhov) [reply src/common/transformations/tests/common_optimizations/attention_mask_shape_replacer.cpp:98]: pretty sure, before SDPAtoPA it is 3d, tests failed because of pattern changes, fixed

File: src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp (hunk 11 above)
Comment (mryzhov) [reply src/common/transformations/src/transformations/paged_attention/attention_mask_shape_replacer.cpp:45]: that how the graph looks right after the PA transformation.

**Review by CuriousPanCake — APPROVED** [review]

File: src/common/transformations/include/transformations/paged_attention/attention_mask_shape_replacer.hpp (hunk 10 above)
Comment (mryzhov) [reply src/common/transformations/include/transformations/paged_attention/attention_mask_shape_replacer.hpp:32]: done
